### PR TITLE
fix: preserve Request body source in patch-fetch

### LINKED ISSUE #90826: Bug: `expected non-null body source` when passing a `Request` object to fetch AND receiving an error response (Node >= 24.14.0)
### Link to the code that reproduces this issue

https://github.com/masakij/nextjs-expected-non-null-body-source-repro

### To Reproduce

The reproduction is very simple. Execute this inside an App Router handler (e.g., Server Action or Route Handler) targeting an endpoint that returns a 4xx or 5xx error:

```ts
// 1. Create a Request object with a body
const req = new Request('https://httpstat.us/401', { // An endpoint that returns 401
  method: 'POST',
  body: JSON.stringify({ test: true }),
  headers: { 'Content-Type': 'application/json' }
});

// 2. Pass the Request object directly to global fetch
// 💥 This crashes immediately on Node >= 24.14.0 when the response is an error
const res = await fetch(req);
```

## Expected Behavior:
The patched `fetch` should return the `Response` object with `res.ok === false` without crashing, just as it does in a pure Node.js environment.

## Actual Behavior:
The application crashes with `TypeError: expected non-null body source at ignore-listed frames`.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 25.3.0: Wed Jan 28 20:53:31 PST 2026; root:xnu-12377.81.4~5/RELEASE_ARM64_T8122
  Available memory (MB): 16384
  Available CPU cores: 8
Binaries:
  Node: 24.14.0
  npm: 11.9.0
  Yarn: N/A
  pnpm: 10.29.3
Relevant Packages:
  next: 16.1.6 // Latest available version is detected (16.1.6).
  eslint-config-next: N/A
  react: 19.2.4
  react-dom: 19.2.4
  typescript: 5.9.3
Next.js Config:
  output: standalone
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local)

## Description:

In the App Router environment, using the global patched fetch crashes with `TypeError: expected non-null body source` under a very specific set of conditions:

You pass a `Request` object (that contains a body, e.g., POST/PUT) directly to `fetch()`.

The request has `Content-Type: application/json`.

The server responds with an error status (e.g., 401 or 500).

This issue surfaces specifically on newer Node.js versions (>= v24.14.0 and probably later v25.x) where undici strictly enforces stream locking.
It appears that when Next.js receives a JSON error response, its internal `patch-fetch` logic attempts to read or clone the original `Request` object (likely for error logging, tracing, or cache evaluation). Because the stream has already been consumed by the actual network request, attempting to access it again throws this error.

This commonly affects users of wrapper libraries like `openapi-fetch` that construct a `Request` object internally and pass it to the global `fetch`.

## Current Workarounds:
Until this is fixed in the framework, developers can use one of the following workarounds depending on their use case:

1. Deconstruct the Request: Pass the URL and `RequestInit` separately instead of the `Request` object.

```ts
// ❌ Fails on error response
fetch(new Request(url, init)); 

// ✅ Works perfectly
fetch(url, init); 
```

2. Use Native Undici Fetch: Bypass the Next.js patched fetch completely for non-GET requests by importing the native fetch: `import { fetch as undiciFetch } from 'undici';`

3. Downgrade Node.js: Use Node.js `<= v24.13.1` where undici's stream handling was less strict.

4. Change Content-Type: If the API allows, using a `Content-Type` other than `application/json` seems to bypass the internal cloning logic causing the crash.

---

_(Note: The investigation and debugging were done manually, but I used an AI assistant to help structure and translate this issue report into English.)_

### PR DESCRIPTION
**Pull request by anthony-schanen**
Fixes #90826
Related #83001

### What?

We should preserve the original Request's internal body source when `patch-fetch` reconstructs the Request for uncached POST requests.

### Why?

When `fetch(new Request(url, { method: 'POST', body: JSON.stringify({}) }))` is called in a Server Action or Route Handler on Node >= 24.14.0, it throws `TypeError: expected non-null body source`.

`patch-fetch.ts` was extracting `reqInput.body` (a ReadableStream) and passing it to `new Request(reqInput.url, { body: stream })`. Node 24.14+ (via undici) validates that the body has an internal "source", ReadableStreams extracted via `.body` don't carry that source.

The existing `_ogBody` mechanism only helps when caching is enabled (it's set during `generateCacheKey()`). For uncached POST requests, `_ogBody` is never set, so the fallback to `reqInput.body` triggers the crash.

### How?

Split the `isRequestInput` block into two branches:

- **if `_ogBody` exists**  the body was consumed by `generateCacheKey()`, Request is disturbed. Reconstruct from URL string with the preserved body (thats the existing behavior).
- **if `_ogBody` is absent**  the body stream is intact. Use `new Request(reqInput, reqOptions)` instead of `new Request(reqInput.url, reqOptions)` to preserve the internal body source. This follows the same pattern as `dedupe-fetch.ts` which already reuses the original Request to avoid disturbing the body.

I also added a unit test verifying uncached POST requests preserve method and non-null body through to `originFetch`. Verified with [reproduction repo](https://github.com/masakij/nextjs-expected-non-null-body-source-repro) on Node 24.14.0.

### CODE REVIEW

**Review by unstubbable — COMMENTED** [review]
Using Node v24.14.0, the test also succeeds for me without your fix.

**Comment by anthony-schanen** [comment]
> Using Node v24.14.0, the test also succeeds for me without your fix.

Thanks for pointing that mistake out. The mocked fetch never went through undici, so the body source validation did not get triggered. I've fixed the test. It now spins up a HTTP server and returns 401 (another mistake in the test). That should match the reproduction from the original issue's reproduction repo. I have the test failing without the fix implemented showing 
`expected non-null body source` and passing with the fix implemented.

Additionally, I am not sure if using `createServer` in a unit test is okay here. I didn't see it used in any other instance within related unit tests, but I could not ideate a way to trigger the undici validation with a mock.

File: packages/next/src/server/lib/patch-fetch.ts [hunk 1]
```diff
@@ -808,15 +803,23 @@ export function createPatchedFetcher(
...
               reqOptions[field] = reqInput[field]
             }
-            input = new Request(reqInput.url, reqOptions)
+
+            const ogBody = (reqInput as any)._ogBody
+            if (ogBody !== undefined) {
+              reqOptions.body = ogBody
+              input = new Request(reqInput.url, reqOptions)
+            } else {
+              if (isStale) {
```
Comment (timneutkens) [inline packages/next/src/server/lib/patch-fetch.ts:822]: This is different than the original implementation. Can you add a comment explaining why this is needed.

**Review by timneutkens — CHANGES_REQUESTED** [review]
Worth adding an e2e test for this instead of a unit test to make sure it's fully covered.

The contributing docs explain how to set it up using `pnpm new-test`

File: packages/next/src/server/lib/patch-fetch.ts (hunk 1 above)
Comment (anthony-schanen) [reply packages/next/src/server/lib/patch-fetch.ts:822]: Added a comment explaining the branch.

**Comment by anthony-schanen** [comment]
Replaced the unit test with an e2e test

**Comment by spencerwalsh651-cmyk** [comment]
The http server return 401

**Comment by anthony-schanen** [comment]
Hey @timneutkens, just wanted to bump this. Please let me know if there's anything else needed on this

**Review by unstubbable — APPROVED** [review]

**Comment by tom-sherman** [comment]
Is it possible for this issue to occur when the server responds with 200? I'm also seeing this issue when using `new Request` but in my case I'm pretty sure the server is responding with an ok.

Also I can't reproduce this locally even with 24.14, only in prod on Vercel.

**Comment by anthony-schanen** [comment]
@unstubbable My commits were not originally signed. I just attempted to sign them now but auto-merge disabled as a result. Just a heads up

**Comment by tom-sherman** [comment]
I found the corresponding undici bug: https://github.com/nodejs/undici/issues/5004#issuecomment-4212579595

Also, I was able to reproduce this on servers that respond with 407 - not just 401. These were the only two status codes I could find that triggered the bug.

**Comment by anthony-schanen** [comment]
> I found the corresponding undici bug: https://github.com/nodejs/undici/issues/5004#issuecomment-4212579595
> 
> 
> 
> Also, I was able to reproduce this on servers that respond with 407 - not just 401. These were the only two status codes I could find that triggered the bug.

I suppose this should be triggered by anything that would cause the source to be consulted when it's null. This probably happens on 4xx errors. As for your 200 scenario I'm not sure. Just speculating though.

**Comment by tom-sherman** [comment]
For my case turns out the server was actually returning 401, I didn't have the correct logging setup to catch this 😅 looks like there's a specific regression in undici for this so this fix only needs to exist while the fix to that regression rolls out. Probably worth adding the fix PR as a code comment? https://github.com/nodejs/undici/pull/4941
